d2/src/main/java/com/linkedin/d2/balancer/clients/RetryClient.java

```diff
@@ -231,12 +296,43 @@ public void onError(Throwable e)
           }
           else
           {
+            double maxClientRequestRetryRatio;
+            _balancer.getLoadBalancedServiceProperties(_serviceName, new Callback<ServiceProperties>() {
+              @Override
+              public void onError(Throwable e) {
+                LOG.warn("Failed to fetch transportClientProperties ", e);
+              }
+
+              @Override
+              public void onSuccess(ServiceProperties result) {
+                _transportClientProperties = result.getTransportClientProperties();
+              }
+            });
+
+            if (_transportClientProperties == null)
```

Comment:
When you get the _transportClientProperties in async manner, you will not immediately obtain the value. After you invoke _balancer.getLoadBalancedServiceProperties asynchronously, I think when you execute to line 312, for sure it will be null, we should notify a callback instead.

Reply:
Fixed. Make isBelowRetryRatio an async call as well.